File: python/kindergarten-garden/kindergarten_garden.py
plant_types = {
  'G': 'Grass', 'C': 'Clover', 'R': 'Radishes', 'V': 'Violets', 
}

class Garden(object):
  def __init__(self, diagram, students = None):
    self.rows = diagram.split('\n')
    self.students = sorted(students) if students else [
      'Alice', 'Bob', 'Charlie', 'David',
      'Eve', 'Fred', 'Ginny', 'Harriet',
      'Ileana', 'Joseph', 'Kincaid', 'Larry',
    ]

  def plants(self, student):
    start = self.students.index(student) * 2
    res = []
    for i in range(2):
      for j in range(start, start+2):
        res.append(plant_types[self.rows[i][j]])
    return res
    # or messy ineficient one-liner:
    # return [plant_types[self.rows[i][j]] for i in range(2) for j in range(self.students.index(student) * 2, (self.students.index(student) * 2)+2)]

File: typescript/kindergarten-garden/kindergarten-garden.ts
const PLANTS: { [letter: string]: string } = {
  V: 'violets',
  C: 'clover',
  G: 'grass',
  R: 'radishes',
}

const defaultStudents: string[] = [
  'Alice', 'Bob', 'Charlie', 'David', 'Eve', 'Fred',
  'Ginny', 'Harriet', 'Ileana', 'Joseph', 'Kincaid', 'Larry'
]

export default class Garden {
  [student: string]: string[]

  constructor(diagram: string, students?: string[]) {
    const garden = diagram.split('\n')
    students = students ? students.sort() : defaultStudents
    students.forEach((student, i) => {
      this[student.toLowerCase()] = [
        PLANTS[garden[0][2 * i]],
        PLANTS[garden[0][2 * i + 1]],
        PLANTS[garden[1][2 * i]],
        PLANTS[garden[1][2 * i + 1]],
      ]
    })
  }
}